### PR TITLE
Toggle put all

### PR DESCRIPTION
**Pull request by Maithem**
Toggle Checkpoint Optimization
Allow the checkpoint writer to enable/disable optimized
checkpoints (i.e. putAll per SMREntry).

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **c5855e9**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #788 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/nGKuV97E4x8XC9owIW6g4ICeu72pHMbN)

**Comment by medhavidhawan** [comment]
We are trying multiple strategies so we might not need this patch at all. Please keep it around just in case.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **0753994**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #788 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/CbXSr9hP0xoH1L5eBeQ1rqCcgH24tTgn)

**Comment by Maithem** [comment]
Based on what I understood from @medhavidhawan , this is no longer needed.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **0728f14**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #788 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/AhmRnuz4g7mVlmZO0OqX8BikJ3vX13EV)

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 1]
```diff
@@ -212,31 +220,57 @@ public long startCheckpoint() {
...
+            }
+        } else {
+            Iterators.partition(map.keySet().stream()
+                    .map(k -> {
+                        return new SMREntry("put",
+                                new Object[]{keyMutator.apply(k), valueMutator.apply(map.get(k))},
+                                serializer);
+                    }).iterator(), batchSize)
+                    .forEachRemaining(entries -> {
+                        MultiSMREntry smrEntries = new MultiSMREntry();
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: MINOR Remove this use of "MultiSMREntry"; it is deprecated. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
MINOR Remove this use of "MultiSMREntry"; it is deprecated. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 2]
```diff
@@ -212,31 +220,57 @@ public long startCheckpoint() {
...
+                        for (int i = 0; i < ((List) entries).size(); i++) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: MINOR Remove this unnecessary cast to "List". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1905)

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 3]
```diff
@@ -212,31 +220,57 @@ public long startCheckpoint() {
...
+                            smrEntries.addTo((SMREntry) ((List) entries).get(i));
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: MINOR Remove this unnecessary cast to "List". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1905)
MINOR Remove this use of "SMREntry"; it is deprecated. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 4]
```diff
@@ -212,31 +220,57 @@ public long startCheckpoint() {
...
-            // for an accurate count of serialized bytes of SRMEntries.
-            numBytes += cp.getSmrEntriesBytes();
+                numEntries++;
+                // CheckpointEntry::serialize() has a side-effect we use
+                // for an accurate count of serialized bytes of SRMEntries.
+                numBytes += cp.getSmrEntriesBytes();
+            }
+        } else {
+            Iterators.partition(map.keySet().stream()
+                    .map(k -> {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: MINOR Remove useless curly braces around statement and then remove useless return keyword [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1602)

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 5]
```diff
@@ -212,31 +220,57 @@ public long startCheckpoint() {
...
+                        return new SMREntry("put",
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: MINOR Remove this use of "SMREntry"; it is deprecated. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **d9a5b72**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #788 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/o021AGRsOij8gNdv4LE7qWBI2QsTApav)

**Pull request merged**